observability/exporting.go

```diff
@@ -0,0 +1,122 @@
+)
+
+// loggingExporter is the interface of logging exporter for gRPC Observability.
+// In future, we might expose this to allow users provide custom exporters. But
+// now, it exists for testing purposes.
+type loggingExporter interface {
+	// EmitGrpcLogRecord writes a gRPC LogRecord to cache without blocking.
+	EmitGrpcLogRecord(*grpclogrecordpb.GrpcLogRecord)
+	// Close flushes all pending data and closes the exporter.
+	Close() error
+}
+
+type cloudLoggingExporter struct {
+	projectID string
+	client    *gcplogging.Client
+	logger    *gcplogging.Logger
+}
+
+func newCloudLoggingExporter(ctx context.Context, projectID string) (*cloudLoggingExporter, error) {
+	c, err := gcplogging.NewClient(ctx, fmt.Sprintf("projects/%v", projectID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cloudLoggingExporter: %v", err)
+	}
+	defer logger.Infof("Successfully created cloudLoggingExporter")
+	customTags := getCustomTags(os.Environ())
+	if len(customTags) != 0 {
+		logger.Infof("Adding custom tags: %+v", customTags)
+	}
+	return &cloudLoggingExporter{
+		projectID: projectID,
+		client:    c,
+		logger:    c.Logger("grpc", gcplogging.CommonLabels(customTags)),
+	}, nil
+}
+
+// mapLogLevelToSeverity maps the gRPC defined log level to Cloud Logging's
+// Severity. The canonical definition can be found at
+// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#LogSeverity.
+var logLevelToSeverity = map[grpclogrecordpb.GrpcLogRecord_LogLevel]gcplogging.Severity{
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_UNKNOWN:  0,
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_TRACE:    100, // Cloud Logging doesn't have a trace level, treated as DEBUG.
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_DEBUG:    100,
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_INFO:     200,
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_WARN:     400,
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_ERROR:    500,
+	grpclogrecordpb.GrpcLogRecord_LOG_LEVEL_CRITICAL: 600,
+}
+
+var protoToJSONOptions = &protojson.MarshalOptions{
+	UseProtoNames:  true,
+	UseEnumNumbers: false,
+}
+
+func (cle *cloudLoggingExporter) EmitGrpcLogRecord(l *grpclogrecordpb.GrpcLogRecord) {
+	// Converts the log record content to a more readable format via protojson.
+	jsonBytes, err := protoToJSONOptions.Marshal(l)
+	if err != nil {
+		logger.Infof("Unable to marshal log record: %v", l)
+		return
+	}
+	var payload map[string]interface{}
+	err = json.Unmarshal(jsonBytes, &payload)
+	if err != nil {
+		logger.Infof("Unable to unmarshal bytes to JSON: %v", jsonBytes)
+		return
+	}
+	entry := gcplogging.Entry{
+		Timestamp: l.Timestamp.AsTime(),
+		Severity:  logLevelToSeverity[l.LogLevel],
+		Payload:   payload,
+	}
+	cle.logger.Log(entry)
+	if logger.V(2) {
+		logger.Infof("Uploading event to CloudLogging: %+v", entry)
+	}
+}
+
+func (cle *cloudLoggingExporter) Close() error {
+	if cle.logger != nil {
+		if err := cle.logger.Flush(); err != nil {
+			return err
```

Comment:
What does Flush error mean to the client? Is the client also dead?
This will skip the cliet.Close() later. Is that OK? Will that leak resources?

Reply:
Or let's combine the errors from Flush and Close?

```go
var errL, errC error
if logger != nil { errL = logger.Flush() }
if client != nil { errL = client.Close() }
if errL == nil && errC == nil { return nil }
// Not sure what to do if both are non-nil.
// Do we know which is more important? Maybe the client close error would override flush error?
// Or just return both.
```

Reply:
Done-ish. PTALAA.

The latest commit explores the idea of combining the errors if both are present.